File: javascript/src/ir-v3-conformance.mjs
import { canonicalHash, canonicalJson } from "./canonical.mjs";
import {
  decodeV3Value,
  encodeV3Value,
  TevScriptError,
} from "./ir-v3-values.mjs";
import { validateProgramIrV3 } from "./ir-v3-validation.mjs";
import { ScriptRuntimeV3 } from "./runtime-v3.mjs";

const STABLE = /^[A-Za-z_][A-Za-z0-9_.:/-]*$/;
const LOCAL = /^[A-Za-z_][A-Za-z0-9_]*$/;
const HASH = /^[0-9a-f]{64}$/;

function fail(code, path, message) {
  throw new TevScriptError(code, `${path}: ${message}`);
}

function objectValue(value, path) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    fail("TEVS_IR_V3_CONFORMANCE_SHAPE", path, "expected object");
  }
  for (const key of Reflect.ownKeys(value)) {
    if (typeof key !== "string") fail("TEVS_IR_V3_CONFORMANCE_SHAPE", path, "expected string object keys");
  }
  return value;
}

function arrayValue(value, path, minimum, maximum) {
  if (!Array.isArray(value) || value.length < minimum || value.length > maximum) {
    fail("TEVS_IR_V3_CONFORMANCE_SHAPE", path, `expected array length in [${minimum}, ${maximum}]`);
  }
  return value;
}

function exactKeys(value, path, expected) {
  const observed = Object.keys(value).sort();
  const target = [...expected].sort();
  if (observed.length !== target.length || observed.some((key, index) => key !== target[index])) {
    const missing = target.filter((key) => !observed.includes(key));
    const extra = observed.filter((key) => !target.includes(key));
    fail(
      "TEVS_IR_V3_CONFORMANCE_SHAPE",
      path,
      `field set mismatch; missing=${JSON.stringify(missing)}, extra=${JSON.stringify(extra)}`,
    );
  }
}

function stableId(value, path) {
  if (typeof value !== "string" || !STABLE.test(value)) {
    fail("TEVS_IR_V3_CONFORMANCE_IDENTIFIER", path, `invalid stable id ${JSON.stringify(value)}`);
  }
  return value;
}

function localId(value, path) {
  if (typeof value !== "string" || !LOCAL.test(value)) {
    fail("TEVS_IR_V3_CONFORMANCE_IDENTIFIER", path, `invalid local id ${JSON.stringify(value)}`);
  }
  return value;
}

function sha256(value, path) {
  if (typeof value !== "string" || !HASH.test(value)) {
    fail("TEVS_IR_V3_CONFORMANCE_HASH", path, "expected lowercase SHA-256");
  }
  return value;
}

function globalCapabilityContracts(ir) {
  const result = new Map();
  for (const entity of ir.entities) {
    for (const capability of entity.capabilities) {
      const contract = {
        parameters: [...capability.parameters],
        returnType: capability.return_type,
        kind: capability.kind,
      };
      const previous = result.get(capability.capability_id);
      if (previous !== undefined && canonicalJson(previous) !== canonicalJson(contract)) {
        fail(
          "TEVS_IR_V3_CONFORMANCE_CAPABILITY_CONFLICT",
          capability.capability_id,
          "program contains conflicting capability contracts",
        );
      }
      result.set(capability.capability_id, contract);
    }
  }
  return result;
}

function validateScriptedCall(call, parameters, returnType, runtime, path) {
  exactKeys(
    call,
    path,
    returnType === "Unit" ? new Set(["arguments"]) : new Set(["arguments", "return"]),
  );
  const arguments_ = arrayValue(call.arguments, `${path}.arguments`, 0, 64);
  if (arguments_.length !== parameters.length) {
    fail(
      "TEVS_IR_V3_CONFORMANCE_CAPABILITY_ARITY",
      path,
      `expected ${parameters.length} arguments, got ${arguments_.length}`,
    );
  }
  parameters.forEach((typeId, index) => {
    const typed = objectValue(arguments_[index], `${path}.arguments[${index}]`);
    exactKeys(typed, `${path}.arguments[${index}]`, new Set(["type", "value"]));
    if (typed.type !== typeId) {
      fail(
        "TEVS_IR_V3_CONFORMANCE_CAPABILITY_TYPE",
        path,
        `argument ${index} expected ${typeId}, got ${JSON.stringify(typed.type)}`,
      );
    }
    decodeV3Value(typeId, typed.value, runtime.typeTable, {
      context: `${path}.arguments[${index}]`,
    });
  });
  if (returnType !== "Unit") {
    const rawReturn = objectValue(call.return, `${path}.return`);
    exactKeys(rawReturn, `${path}.return`, new Set(["type", "value"]));
    if (rawReturn.type !== returnType) {
      fail(
        "TEVS_IR_V3_CONFORMANCE_CAPABILITY_RETURN_TYPE",
        path,
        `expected return ${returnType}, got ${JSON.stringify(rawReturn.type)}`,
      );
    }
    decodeV3Value(returnType, rawReturn.value, runtime.typeTable, {
      context: `${path}.return`,
    });
  }
}

class ScenarioHost {
  constructor(runtime, scenarioScripts) {
    this.runtime = runtime;
    this.contracts = globalCapabilityContracts(runtime.ir);
    this.scripts = new Map();
    this.transcript = [];
    let previousId = null;

    scenarioScripts.forEach((raw, index) => {
      const path = `$.capabilities[${index}]`;
      const item = objectValue(raw, path);
      exactKeys(item, path, new Set(["capability_id", "calls"]));
      const capabilityId = stableId(item.capability_id, `${path}.capability_id`);
      if (this.scripts.has(capabilityId)) {
        fail("TEVS_IR_V3_CONFORMANCE_CAPABILITY_DUPLICATE", path, `duplicate capability script ${JSON.stringify(capabilityId)}`);
      }
      if (previousId !== null && capabilityId <= previousId) {
        fail("TEVS_IR_V3_CONFORMANCE_CAPABILITY_ORDER", path, "capability scripts must be strictly sorted by id");
      }
      previousId = capabilityId;
      const contract = this.contracts.get(capabilityId);
      if (contract === undefined) {
        fail("TEVS_IR_V3_CONFORMANCE_CAPABILITY_UNKNOWN", path, `scenario scripts undeclared capability ${JSON.stringify(capabilityId)}`);
      }
      const calls = arrayValue(item.calls, `${path}.calls`, 0, 4096);
      calls.forEach((call, callIndex) => validateScriptedCall(
        objectValue(call, `${path}.calls[${callIndex}]`),
        contract.parameters,
        contract.returnType,
        runtime,
        `${path}.calls[${callIndex}]`,
      ));
      this.scripts.set(capabilityId, {
        capabilityId,
        parameters: contract.parameters,
        returnType: contract.returnType,
        calls,
        cursor: 0,
      });
    });
  }

  bindings() {
    return Object.fromEntries(
      [...this.scripts].map(([capabilityId, script]) => [
        capabilityId,
        (...arguments_) => this.invokeScript(script, arguments_),
      ]),
    );
  }

  invokeScript(script, arguments_) {
    if (script.cursor >= script.calls.length) {
      fail(
        "TEVS_IR_V3_CONFORMANCE_CAPABILITY_CALL_OVERFLOW",
        script.capabilityId,
        "runtime invoked capability more times than scripted",
      );
    }
    const expected = script.calls[script.cursor];
    const callIndex = this.transcript.length;
    const encodedArguments = script.parameters.map((typeId, index) => ({
      type: typeId,
      value: encodeV3Value(
        typeId,
        arguments_[index],
        this.runtime.typeTable,
        { context: `capability ${script.capabilityId} argument ${index}` },
      ),
    }));
    if (canonicalJson(encodedArguments) !== canonicalJson(expected.arguments)) {
      fail(
        "TEVS_IR_V3_CONFORMANCE_CAPABILITY_ARGUMENTS",
        script.capabilityId,
        `call ${script.cursor} arguments differ from scripted canonical values`,
      );
    }
    const transcriptEntry = {
      index: callIndex,
      capability_id: script.capabilityId,
      arguments: encodedArguments,
    };
    script.cursor += 1;
    if (script.returnType === "Unit") {
      if (Object.hasOwn(expected, "return")) {
        fail(
          "TEVS_IR_V3_CONFORMANCE_CAPABILITY_RETURN",
          script.capabilityId,
          "Unit scripted call must not contain return",
        );
      }
      this.transcript.push(transcriptEntry);
      return undefined;
    }
    const rawReturn = objectValue(expected.return, `capability ${script.capabilityId}.return`);
    transcriptEntry.return = rawReturn;
    this.transcript.push(transcriptEntry);
    return decodeV3Value(
      script.returnType,
      rawReturn.value,
      this.runtime.typeTable,
      { context: `capability ${script.capabilityId} scripted return` },
    );
  }

  verifyConsumed() {
    const pending = Object.fromEntries(
      [...this.scripts]
        .filter(([, script]) => script.cursor !== script.calls.length)
        .map(([capabilityId, script]) => [capabilityId, script.calls.length - script.cursor]),
    );
    if (Object.keys(pending).length !== 0) {
      fail(
        "TEVS_IR_V3_CONFORMANCE_CAPABILITY_CALL_UNDERFLOW",
        "$.capabilities",
        `scripted capability calls not consumed exactly: ${canonicalJson(pending)}`,
      );
    }
  }
}

function stateWitness(runtime) {
  return [...runtime.entities.keys()].sort().map((entityId) => {
    const entity = runtime.entities.get(entityId);
    const state = Object.fromEntries(
      [...entity.state.keys()].sort().map((stateName) => [
        stateName,
        {
          type: entity.stateTypes.get(stateName),
          value: encodeV3Value(
            entity.stateTypes.get(stateName),
            entity.state.get(stateName),
            runtime.typeTable,
            { context: `state witness ${entityId}.${stateName}` },
          ),
        },
      ]),
    );
    return { entity_id: entityId, state };
  });
}

function eventWitness(event, runtime) {
  return {
    entity_id: event.entity_id,
    event_id: event.event_id,
    arguments: event.argument_types.map((typeId, index) => ({
      type: typeId,
      value: encodeV3Value(
        typeId,
        event.arguments[index],
        runtime.typeTable,
        { context: `event witness ${event.event_id}[${index}]` },
      ),
    })),
  };
}

function decodeInvocationArguments(runtime, entityId, eventId, arguments_) {
  const entity = runtime.entities.get(entityId);
  if (!entity) fail("TEVS_IR_V3_CONFORMANCE_ENTITY", entityId, "unknown scenario entity");
  const handler = entity.handlers.get(eventId);
  let expected;
  if (handler !== undefined) expected = handler.parameters.map((item) => item.type);
  else {
    expected = entity.emittedEventTypes.get(eventId);
    if (expected === undefined) {
      if (arguments_.length !== 0) {
        fail(
          "TEVS_IR_V3_CONFORMANCE_EVENT_SIGNATURE",
          eventId,
          "unhandled scenario event with arguments has no portable signature",
        );
      }
      expected = [];
    }
  }
  if (expected.length !== arguments_.length) {
    fail(
      "TEVS_IR_V3_CONFORMANCE_EVENT_ARITY",
      eventId,
      `expected ${expected.length} arguments, got ${arguments_.length}`,
    );
  }
  return expected.map((expectedType, index) => {
    const typed = objectValue(arguments_[index], `scenario argument ${index}`);
    exactKeys(typed, `scenario argument ${index}`, new Set(["type", "value"]));
    if (typed.type !== expectedType) {
      fail(
        "TEVS_IR_V3_CONFORMANCE_EVENT_TYPE",
        eventId,
        `argument ${index} expected ${expectedType}, got ${JSON.stringify(typed.type)}`,
      );
    }
    return decodeV3Value(expectedType, typed.value, runtime.typeTable, {
      context: `scenario ${entityId}.${eventId}[${index}]`,
    });
  });
}

function validateScenario(scenario, program) {
  exactKeys(
    scenario,
    "$",
    new Set([
      "schema", "scenario_id", "program_semantic_hash",
      "source_semantic_hash", "capabilities", "steps",
    ]),
  );
  if (scenario.schema !== "TEV_SCRIPT_IR_V3_SCENARIO_V1") {
    fail("TEVS_IR_V3_CONFORMANCE_SCENARIO_SCHEMA", "$.schema", "unexpected scenario schema");
  }
  stableId(scenario.scenario_id, "$.scenario_id");
  const programHash = sha256(scenario.program_semantic_hash, "$.program_semantic_hash");
  const sourceHash = sha256(scenario.source_semantic_hash, "$.source_semantic_hash");
  if (programHash !== program.semantic_hash) {
    fail("TEVS_IR_V3_CONFORMANCE_PROGRAM_HASH", "$.program_semantic_hash", "scenario targets a different IR V3 semantic hash");
  }
  if (sourceHash !== program.source_semantic_hash) {
    fail("TEVS_IR_V3_CONFORMANCE_SOURCE_HASH", "$.source_semantic_hash", "scenario targets a different source semantic hash");
  }
  arrayValue(scenario.capabilities, "$.capabilities", 0, 8192);
  const steps = arrayValue(scenario.steps, "$.steps", 0, 1024);
  steps.forEach((raw, index) => {
    const step = objectValue(raw, `$.steps[${index}]`);
    exactKeys(step, `$.steps[${index}]`, new Set(["entity_id", "event_id", "arguments"]));
    localId(step.entity_id, `$.steps[${index}].entity_id`);
    localId(step.event_id, `$.steps[${index}].event_id`);
    arrayValue(step.arguments, `$.steps[${index}].arguments`, 0, 64);
  });
  return scenario;
}

export function runIrV3Conformance(ir, scenario) {
  const program = structuredClone(ir);
  validateProgramIrV3(program);
  const scenarioObject = validateScenario(structuredClone(scenario), program);
  const scenarioHash = canonicalHash(scenarioObject);

  const bootstrap = new ScriptRuntimeV3(program, {}, {
    expectedSourceSemanticHash: scenarioObject.source_semantic_hash,
  });
  const host = new ScenarioHost(
    bootstrap,
    arrayValue(scenarioObject.capabilities, "$.capabilities", 0, 8192),
  );
  const runtime = new ScriptRuntimeV3(program, host.bindings(), {
    expectedSourceSemanticHash: scenarioObject.source_semantic_hash,
  });
  host.runtime = runtime;

  const initialState = stateWitness(runtime);
  const initialStateHash = canonicalHash(initialState);
  const steps = [];
  scenarioObject.steps.forEach((rawStep, index) => {
    const entityId = rawStep.entity_id;
    const eventId = rawStep.event_id;
    const arguments_ = decodeInvocationArguments(runtime, entityId, eventId, rawStep.arguments);
    const emitted = runtime.invoke(entityId, eventId, ...arguments_);
    const state = stateWitness(runtime);
    steps.push({
      index,
      entity_id: entityId,
      event_id: eventId,
      emitted: emitted.map((event) => eventWitness(event, runtime)),
      state_hash: canonicalHash(state),
    });
  });

  host.verifyConsumed();
  const finalState = stateWitness(runtime);
  const finalStateHash = canonicalHash(finalState);
  const body = {
    schema: "TEV_SCRIPT_IR_V3_CONFORMANCE_RECEIPT_V1",
    scenario_id: scenarioObject.scenario_id,
    scenario_hash: scenarioHash,
    program_semantic_hash: program.semantic_hash,
    source_semantic_hash: program.source_semantic_hash,
    initial_state_hash: initialStateHash,
    steps,
    capability_calls: host.transcript,
    final_state: finalState,
    final_state_hash: finalStateHash,
  };
  const receiptHash = canonicalHash(body);
  const receipt = { ...body, receipt_hash: receiptHash };
  return {
    receipt,
    canonicalJson: canonicalJson(receipt),
    receiptHash,
  };
}

File: tev_script/ir_v3_conformance.py
from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Any, Mapping

from .canonical import canonical_hash, canonical_json
from .diagnostics import TevScriptError
from .ir_v3_validation import validate_program_ir_v3
from .ir_v3_values import decode_v3_value, encode_v3_value
from .runtime_v3 import EmittedEventV3, ScriptRuntimeV3

_STABLE = re.compile(r"^[A-Za-z_][A-Za-z0-9_.:/-]*$")
_LOCAL = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")
_HASH = re.compile(r"^[0-9a-f]{64}$")


@dataclass(frozen=True, slots=True)
class IrV3ConformanceReceiptBundle:
    receipt: dict[str, object]
    canonical_json: str
    receipt_hash: str


@dataclass(slots=True)
class _CapabilityScript:
    capability_id: str
    parameters: tuple[str, ...]
    return_type: str
    calls: list[dict[str, object]]
    cursor: int = 0


class _ScenarioHost:
    def __init__(
        self,
        runtime: ScriptRuntimeV3,
        scenario_scripts: list[dict[str, object]],
    ) -> None:
        self.runtime = runtime
        self.contracts = _global_capability_contracts(runtime.ir)
        self.scripts: dict[str, _CapabilityScript] = {}
        self.transcript: list[dict[str, object]] = []
        previous_id: str | None = None
        for index, raw in enumerate(scenario_scripts):
            path = f"$.capabilities[{index}]"
            item = _object(raw, path)
            _exact_keys(item, path, {"capability_id", "calls"})
            capability_id = _stable(item["capability_id"], path + ".capability_id")
            if capability_id in self.scripts:
                _fail("TEVS_IR_V3_CONFORMANCE_CAPABILITY_DUPLICATE", path, f"duplicate capability script {capability_id!r}")
            if previous_id is not None and capability_id <= previous_id:
                _fail("TEVS_IR_V3_CONFORMANCE_CAPABILITY_ORDER", path, "capability scripts must be strictly sorted by id")
            previous_id = capability_id
            contract = self.contracts.get(capability_id)
            if contract is None:
                _fail("TEVS_IR_V3_CONFORMANCE_CAPABILITY_UNKNOWN", path, f"scenario scripts undeclared capability {capability_id!r}")
            parameters, return_type, _kind = contract
            calls = _array(item["calls"], path + ".calls", 0, 4096)
            for call_index, call in enumerate(calls):
                _validate_scripted_call(
                    _object(call, f"{path}.calls[{call_index}]"),
                    parameters,
                    return_type,
                    runtime,
                    f"{path}.calls[{call_index}]",
                )
            self.scripts[capability_id] = _CapabilityScript(
                capability_id,
                parameters,
                return_type,
                calls,
            )

    def bindings(self) -> dict[str, Any]:
        return {
            capability_id: self._provider(script)
            for capability_id, script in self.scripts.items()
        }

    def _provider(self, script: _CapabilityScript):
        def provider(*arguments: Any) -> Any:
            if script.cursor >= len(script.calls):
                _fail(
                    "TEVS_IR_V3_CONFORMANCE_CAPABILITY_CALL_OVERFLOW",
                    script.capability_id,
                    "runtime invoked capability more times than scripted",
                )
            expected = script.calls[script.cursor]
            call_index = len(self.transcript)
            encoded_arguments = [
                {
                    "type": type_id,
                    "value": encode_v3_value(
                        type_id,
                        value,
                        self.runtime.type_table,
                        context=f"capability {script.capability_id} argument {index}",
                    ),
                }
                for index, (type_id, value) in enumerate(
                    zip(script.parameters, arguments, strict=True)
                )
            ]
            if canonical_json(encoded_arguments) != canonical_json(expected["arguments"]):
                _fail(
                    "TEVS_IR_V3_CONFORMANCE_CAPABILITY_ARGUMENTS",
                    script.capability_id,
                    f"call {script.cursor} arguments differ from scripted canonical values",
                )
            transcript_entry: dict[str, object] = {
                "index": call_index,
                "capability_id": script.capability_id,
                "arguments": encoded_arguments,
            }
            script.cursor += 1
            if script.return_type == "Unit":
                if "return" in expected:
                    _fail(
                        "TEVS_IR_V3_CONFORMANCE_CAPABILITY_RETURN",
                        script.capability_id,
                        "Unit scripted call must not contain return",
                    )
                self.transcript.append(transcript_entry)
                return None

            raw_return = expected.get("return")
            if raw_return is None:
                _fail(
                    "TEVS_IR_V3_CONFORMANCE_CAPABILITY_RETURN",
                    script.capability_id,
                    "non-Unit scripted call requires return",
                )
            typed_return = _object(raw_return, f"capability {script.capability_id}.return")
            transcript_entry["return"] = typed_return
            self.transcript.append(transcript_entry)
            return decode_v3_value(
                script.return_type,
                typed_return["value"],
                self.runtime.type_table,
                context=f"capability {script.capability_id} scripted return",
            )

        return provider

    def verify_consumed(self) -> None:
        pending = {
            capability_id: len(script.calls) - script.cursor
            for capability_id, script in self.scripts.items()
            if script.cursor != len(script.calls)
        }
        if pending:
            _fail(
                "TEVS_IR_V3_CONFORMANCE_CAPABILITY_CALL_UNDERFLOW",
                "$.capabilities",
                f"scripted capability calls not consumed exactly: {pending}",
            )


def run_ir_v3_conformance(
    ir: Mapping[str, Any],
    scenario: Mapping[str, Any],
) -> IrV3ConformanceReceiptBundle:
    program = dict(ir)
    validate_program_ir_v3(program)
    scenario_obj = _validate_scenario(dict(scenario), program)
    scenario_hash = canonical_hash(scenario_obj)

    bootstrap = ScriptRuntimeV3(
        program,
        expected_source_semantic_hash=str(scenario_obj["source_semantic_hash"]),
    )
    host = _ScenarioHost(
        bootstrap,
        _array(scenario_obj["capabilities"], "$.capabilities", 0, 8192),
    )
    # Recreate the runtime with the fully validated scripted capability bindings.
    runtime = ScriptRuntimeV3(
        program,
        host.bindings(),
        expected_source_semantic_hash=str(scenario_obj["source_semantic_hash"]),
    )
    host.runtime = runtime

    initial_state = _state_witness(runtime)
    initial_state_hash = canonical_hash(initial_state)
    step_receipts: list[dict[str, object]] = []
    steps = _array(scenario_obj["steps"], "$.steps", 0, 1024)
    for index, raw_step in enumerate(steps):
        step = _object(raw_step, f"$.steps[{index}]")
        entity_id = str(step["entity_id"])
        event_id = str(step["event_id"])
        arguments = _decode_invocation_arguments(
            runtime,
            entity_id,
            event_id,
            _array(step["arguments"], f"$.steps[{index}].arguments", 0, 64),
        )
        emitted = runtime.invoke(entity_id, event_id, *arguments)
        state = _state_witness(runtime)
        step_receipts.append(
            {
                "index": index,
                "entity_id": entity_id,
                "event_id": event_id,
                "emitted": [_event_witness(event, runtime) for event in emitted],
                "state_hash": canonical_hash(state),
            }
        )

    host.verify_consumed()
    final_state = _state_witness(runtime)
    final_state_hash = canonical_hash(final_state)
    body: dict[str, object] = {
        "schema": "TEV_SCRIPT_IR_V3_CONFORMANCE_RECEIPT_V1",
        "scenario_id": str(scenario_obj["scenario_id"]),
        "scenario_hash": scenario_hash,
        "program_semantic_hash": str(program["semantic_hash"]),
        "source_semantic_hash": str(program["source_semantic_hash"]),
        "initial_state_hash": initial_state_hash,
        "steps": step_receipts,
        "capability_calls": host.transcript,
        "final_state": final_state,
        "final_state_hash": final_state_hash,
    }
    receipt_hash = canonical_hash(body)
    receipt = dict(body)
    receipt["receipt_hash"] = receipt_hash
    return IrV3ConformanceReceiptBundle(
        receipt,
        canonical_json(receipt),
        receipt_hash,
    )


def _state_witness(runtime: ScriptRuntimeV3) -> list[dict[str, object]]:
    result: list[dict[str, object]] = []
    for entity_id in sorted(runtime.entities):
        entity = runtime.entities[entity_id]
        state = {
            state_name: {
                "type": entity.state_types[state_name],
                "value": encode_v3_value(
                    entity.state_types[state_name],
                    entity.state[state_name],
                    runtime.type_table,
                    context=f"state witness {entity_id}.{state_name}",
                ),
            }
            for state_name in sorted(entity.state)
        }
        result.append({"entity_id": entity_id, "state": state})
    return result


def _event_witness(event: EmittedEventV3, runtime: ScriptRuntimeV3) -> dict[str, object]:
    return {
        "entity_id": event.entity_id,
        "event_id": event.event_id,
        "arguments": [
            {
                "type": type_id,
                "value": encode_v3_value(
                    type_id,
                    value,
                    runtime.type_table,
                    context=f"event witness {event.event_id}[{index}]",
                ),
            }
            for index, (type_id, value) in enumerate(
                zip(event.argument_types, event.arguments, strict=True)
            )
        ],
    }


def _decode_invocation_arguments(
    runtime: ScriptRuntimeV3,
    entity_id: str,
    event_id: str,
    arguments: list[dict[str, object]],
) -> tuple[Any, ...]:
    entity = runtime.entities.get(entity_id)
    if entity is None:
        _fail("TEVS_IR_V3_CONFORMANCE_ENTITY", entity_id, "unknown scenario entity")
    handler = entity.handlers.get(event_id)
    if handler is not None:
        expected = tuple(str(item["type"]) for item in handler["parameters"])
    else:
        expected = entity.emitted_event_types.get(event_id)
        if expected is None:
            if arguments:
                _fail(
                    "TEVS_IR_V3_CONFORMANCE_EVENT_SIGNATURE",
                    event_id,
                    "unhandled scenario event with arguments has no portable signature",
                )
            expected = ()
    if len(expected) != len(arguments):
        _fail(
            "TEVS_IR_V3_CONFORMANCE_EVENT_ARITY",
            event_id,
            f"expected {len(expected)} arguments, got {len(arguments)}",
        )
    decoded: list[Any] = []
    for index, (expected_type, raw_argument) in enumerate(zip(expected, arguments, strict=True)):
        typed = _object(raw_argument, f"scenario argument {index}")
        _exact_keys(typed, f"scenario argument {index}", {"type", "value"})
        if typed["type"] != expected_type:
            _fail(
                "TEVS_IR_V3_CONFORMANCE_EVENT_TYPE",
                event_id,
                f"argument {index} expected {expected_type}, got {typed['type']!r}",
            )
        decoded.append(
            decode_v3_value(
                expected_type,
                typed["value"],
                runtime.type_table,
                context=f"scenario {entity_id}.{event_id}[{index}]",
            )
        )
    return tuple(decoded)


def _validate_scenario(
    scenario: dict[str, object],
    program: dict[str, object],
) -> dict[str, object]:
    _exact_keys(
        scenario,
        "$",
        {
            "schema", "scenario_id", "program_semantic_hash",
            "source_semantic_hash", "capabilities", "steps",
        },
    )
    if scenario["schema"] != "TEV_SCRIPT_IR_V3_SCENARIO_V1":
        _fail("TEVS_IR_V3_CONFORMANCE_SCENARIO_SCHEMA", "$.schema", "unexpected scenario schema")
    scenario_id = _stable(scenario["scenario_id"], "$.scenario_id")
    del scenario_id
    program_hash = _sha(scenario["program_semantic_hash"], "$.program_semantic_hash")
    source_hash = _sha(scenario["source_semantic_hash"], "$.source_semantic_hash")
    if program_hash != program["semantic_hash"]:
        _fail("TEVS_IR_V3_CONFORMANCE_PROGRAM_HASH", "$.program_semantic_hash", "scenario targets a different IR V3 semantic hash")
    if source_hash != program["source_semantic_hash"]:
        _fail("TEVS_IR_V3_CONFORMANCE_SOURCE_HASH", "$.source_semantic_hash", "scenario targets a different source semantic hash")
    _array(scenario["capabilities"], "$.capabilities", 0, 8192)
    steps = _array(scenario["steps"], "$.steps", 0, 1024)
    for index, raw in enumerate(steps):
        step = _object(raw, f"$.steps[{index}]")
        _exact_keys(step, f"$.steps[{index}]", {"entity_id", "event_id", "arguments"})
        _local(step["entity_id"], f"$.steps[{index}].entity_id")
        _local(step["event_id"], f"$.steps[{index}].event_id")
        _array(step["arguments"], f"$.steps[{index}].arguments", 0, 64)
    return scenario


def _global_capability_contracts(
    ir: Mapping[str, Any],
) -> dict[str, tuple[tuple[str, ...], str, str]]:
    result: dict[str, tuple[tuple[str, ...], str, str]] = {}
    for entity in ir["entities"]:
        for capability in entity["capabilities"]:
            capability_id = str(capability["capability_id"])
            contract = (
                tuple(str(item) for item in capability["parameters"]),
                str(capability["return_type"]),
                str(capability["kind"]),
            )
            previous = result.get(capability_id)
            if previous is not None and previous != contract:
                _fail(
                    "TEVS_IR_V3_CONFORMANCE_CAPABILITY_CONFLICT",
                    capability_id,
                    "program contains conflicting capability contracts",
                )
            result[capability_id] = contract
    return result


def _validate_scripted_call(
    call: dict[str, object],
    parameters: tuple[str, ...],
    return_type: str,
    runtime: ScriptRuntimeV3,
    path: str,
) -> None:
    allowed = {"arguments"} if return_type == "Unit" else {"arguments", "return"}
    _exact_keys(call, path, allowed)
    arguments = _array(call["arguments"], path + ".arguments", 0, 64)
    if len(arguments) != len(parameters):
        _fail(
            "TEVS_IR_V3_CONFORMANCE_CAPABILITY_ARITY",
            path,
            f"expected {len(parameters)} arguments, got {len(arguments)}",
        )
    for index, (type_id, raw) in enumerate(zip(parameters, arguments, strict=True)):
        typed = _object(raw, f"{path}.arguments[{index}]")
        _exact_keys(typed, f"{path}.arguments[{index}]", {"type", "value"})
        if typed["type"] != type_id:
            _fail(
                "TEVS_IR_V3_CONFORMANCE_CAPABILITY_TYPE",
                path,
                f"argument {index} expected {type_id}, got {typed['type']!r}",
            )
        decode_v3_value(
            type_id,
            typed["value"],
            runtime.type_table,
            context=f"{path}.arguments[{index}]",
        )
    if return_type != "Unit":
        raw_return = _object(call["return"], path + ".return")
        _exact_keys(raw_return, path + ".return", {"type", "value"})
        if raw_return["type"] != return_type:
            _fail(
                "TEVS_IR_V3_CONFORMANCE_CAPABILITY_RETURN_TYPE",
                path,
                f"expected return {return_type}, got {raw_return['type']!r}",
            )
        decode_v3_value(
            return_type,
            raw_return["value"],
            runtime.type_table,
            context=path + ".return",
        )


def _fail(code: str, path: str, message: str) -> None:
    raise TevScriptError(code, f"{path}: {message}")


def _object(value: Any, path: str) -> dict[str, Any]:
    if not isinstance(value, dict) or not all(isinstance(key, str) for key in value):
        _fail("TEVS_IR_V3_CONFORMANCE_SHAPE", path, "expected object")
    return value


def _array(value: Any, path: str, minimum: int, maximum: int) -> list[Any]:
    if not isinstance(value, list) or not minimum <= len(value) <= maximum:
        _fail("TEVS_IR_V3_CONFORMANCE_SHAPE", path, f"expected array length in [{minimum}, {maximum}]")
    return value


def _exact_keys(value: Mapping[str, Any], path: str, expected: set[str]) -> None:
    observed = set(value)
    if observed != expected:
        _fail(
            "TEVS_IR_V3_CONFORMANCE_SHAPE",
            path,
            f"field set mismatch; missing={sorted(expected-observed)}, extra={sorted(observed-expected)}",
        )


def _stable(value: Any, path: str) -> str:
    if not isinstance(value, str) or _STABLE.fullmatch(value) is None:
        _fail("TEVS_IR_V3_CONFORMANCE_IDENTIFIER", path, f"invalid stable id {value!r}")
    return value


def _local(value: Any, path: str) -> str:
    if not isinstance(value, str) or _LOCAL.fullmatch(value) is None:
        _fail("TEVS_IR_V3_CONFORMANCE_IDENTIFIER", path, f"invalid local id {value!r}")
    return value


def _sha(value: Any, path: str) -> str:
    if not isinstance(value, str) or _HASH.fullmatch(value) is None:
        _fail("TEVS_IR_V3_CONFORMANCE_HASH", path, "expected lowercase SHA-256")
    return value
